packages/react-query/src/internals/getQueryKey.ts

```diff
@@ -1,11 +1,98 @@
+import {
+  AnyMutationProcedure,
+  AnyQueryProcedure,
+  AnyRouter,
+  DeepPartial,
+  inferProcedureInput,
+} from '@trpc/server';
+import { QueryType, getArrayQueryKey } from '../internals/getArrayQueryKey';
+import { DecorateProcedure, DecoratedProcedureRecord } from '../shared';
+
 /**
  * We treat `undefined` as an input the same as omitting an `input`
  * https://github.com/trpc/trpc/issues/2290
  */
-export function getQueryKey(
+export function getQueryKeyInternal(
   path: string,
   input: unknown,
 ): [string] | [string, unknown] {
   if (path.length) return input === undefined ? [path] : [path, input];
   return [] as unknown as [string];
 }
+
+type GetInfinitQueryInput<TProcedureOrRouter extends AnyQueryProcedure, 
+TInputWithoutCursor = Omit<
+    inferProcedureInput<TProcedureOrRouter>,
+    'cursor'
+  >> =  keyof TInputWithoutCursor extends never
+    ? undefined
+    : DeepPartial<TInputWithoutCursor> | undefined;
+
+type GetQueryProcedureInput<TProcedureOrRouter extends AnyQueryProcedure> =
+  inferProcedureInput<TProcedureOrRouter> extends {
+    cursor?: any;
+  }
+    ? GetInfinitQueryInput<TProcedureOrRouter>
+    : DeepPartial<inferProcedureInput<TProcedureOrRouter>> | undefined;
+
+type GetQueryParams<TProcedureOrRouter extends AnyQueryProcedure> =
+  inferProcedureInput<TProcedureOrRouter> extends undefined
+    ? []
+    : [input?: GetQueryProcedureInput<TProcedureOrRouter>, type?: QueryType];
```

Comment:
Little bit improved version that may improve ts infer times by not repeating infer lookups :

```suggestion

type GetInfinitQueryInput<TProcedureInput, 
TInputWithoutCursor = Omit<
    TProcedureInput,
    'cursor'
  >> =  keyof TInputWithoutCursor extends never
    ? undefined
    : DeepPartial<TInputWithoutCursor> | undefined;

type GetQueryProcedureInput<TProcedureInput> =
  TProcedureInput extends {
    cursor?: any;
  }
    ? GetInfinitQueryInput<TProcedureInput>
    : DeepPartial<TProcedureInput> | undefined;

type GetQueryParams<
          TProcedureOrRouter extends AnyQueryProcedure,
          TProcedureInput = inferProcedureInput<TProcedureOrRouter>
        > =
   TProcedureInput extends undefined
    ? []
    : [input?: GetQueryProcedureInput<TProcedureInput>, type?: QueryType];
```